src/OrleansGoogleUtils/Providers/Streams/GooglePubSubAdapterReceiver.cs

```diff
@@ -0,0 +1,150 @@
+    {
+        private readonly SerializationManager _serializationManager;
+        private GooglePubSubDataManager _pubSub;
+        private long _lastReadMessage;
+        private Task _outstandingTask;
+        private readonly Logger _logger;
+        private readonly IGooglePubSubDataAdapter _dataAdapter;
+        private readonly List<PendingDelivery> _pending;
+
+        public QueueId Id { get; }
+
+        public static IQueueAdapterReceiver Create(SerializationManager serializationManager, QueueId queueId, string projectId, string topicId,
+            string deploymentId, IGooglePubSubDataAdapter dataAdapter, TimeSpan? deadline = null)
+        {
+            if (queueId == null) throw new ArgumentNullException(nameof(queueId));
+            if (dataAdapter == null) throw new ArgumentNullException(nameof(dataAdapter));
+            if (serializationManager == null) throw new ArgumentNullException(nameof(serializationManager));
+
+            var pubSub = new GooglePubSubDataManager(projectId, topicId, queueId.ToString(), deploymentId, deadline);
+            return new GooglePubSubAdapterReceiver(serializationManager, queueId, pubSub, dataAdapter);
+        }
+
+        private GooglePubSubAdapterReceiver(SerializationManager serializationManager, QueueId queueId, GooglePubSubDataManager pubSub, IGooglePubSubDataAdapter dataAdapter)
+        {
+            Id = queueId ?? throw new ArgumentNullException(nameof(queueId));
+            _serializationManager = serializationManager;
+            _pubSub = pubSub ?? throw new ArgumentNullException(nameof(pubSub));
+            _dataAdapter = dataAdapter ?? throw new ArgumentNullException(nameof(dataAdapter));
+            _logger = LogManager.GetLogger(GetType().Name, LoggerType.Provider);
+            _pending = new List<PendingDelivery>();
+        }
+
+        public Task Initialize(TimeSpan timeout)
+        {
+            if (_pubSub != null) return _pubSub.Initialize();
+
+            return Task.CompletedTask;
+        }
+
+        public async Task Shutdown(TimeSpan timeout)
+        {
+            try
+            {
+                // await the last pending operation, so after we shutdown and stop this receiver we don't get async operation completions from pending operations.
+                if (_outstandingTask != null)
+                    await _outstandingTask;
+            }
+            finally
+            {
+                // remember that we shut down so we never try to read from the queue again.
+                _pubSub = null;
+            }
+        }
+
+        public async Task<IList<IBatchContainer>> GetQueueMessagesAsync(int maxCount)
+        {
+            try
+            {
+                var pubSubRef = _pubSub; // store direct ref, in case we are somehow asked to shutdown while we are receiving.    
+                if (pubSubRef == null) return new List<IBatchContainer>();
+
+                int count = maxCount < 0 ? 10 : maxCount;
```

Comment:
why do we choose 10 here? seems like a small number for queueing messages. 
And keep in mind , maxCount can be QueueAdapterConstants.UNLIMITED_GET_QUEUE_MSG, which is a below-zero value "-1"